src/services/Storage.tsx

```diff
@@ -0,0 +1,180 @@
+ * is stored yet on the device.
+ * It is also useful for testing, to mock the actual storage implementation.
+ */
+export const newDefaultStorage = (): Storage => ({
+  wordNodeCards: [],
+  isTrackingEnabled: true,
+  selectedProfessions: null,
+  isDevModeEnabled: false,
+  progress: {},
+  cmsUrlOverwrite: null,
+  customDisciplines: [],
+  userVocabulary: [],
+  nextUserVocabularyId: 1,
+  favorites: [],
+})
+const defaultStorage = newDefaultStorage()
+
+// eslint-disable-next-line consistent-return
+export const getStorageKey = (key: keyof Storage): string => {
+  switch (key) {
+    case 'wordNodeCards':
+      return 'wordNodeCards'
+    case 'isTrackingEnabled':
+      return 'sentryTracking'
+    case 'selectedProfessions':
+      return 'selectedProfessions'
+    case 'isDevModeEnabled':
+      return 'devmode'
+    case 'progress':
+      return 'progress'
+    case 'cmsUrlOverwrite':
+      return 'cms'
+    case 'customDisciplines':
+      return 'customDisciplines'
+    case 'userVocabulary':
+      return 'userVocabulary'
+    case 'nextUserVocabularyId':
+      return 'userVocabularyNextId'
+    case 'favorites':
+      return 'favorites-2'
+  }
+}
+
+export const getStorageItemOr = async <T,>(key: string, defaultValue: T): Promise<T> => {
+  const value = await AsyncStorage.getItem(key)
+  return value ? JSON.parse(value) : defaultValue
+}
+
+export const getStorageItem = async <T extends keyof Storage>(key: T): Promise<Storage[T]> =>
+  getStorageItemOr(getStorageKey(key), defaultStorage[key])
+
+const setStorageItem = async <T extends keyof Storage>(key: T, value: Storage[T]): Promise<void> => {
+  await AsyncStorage.setItem(getStorageKey(key), JSON.stringify(value))
+}
+
+// https://github.com/react-native-async-storage/async-storage/issues/401#issuecomment-2508924008
+export class StorageCache {
+  private readonly listeners: Map<string, Set<() => void>> = new Map()
+  private readonly storage: Storage
+
+  private constructor(storage: Storage) {
+    this.storage = storage
+  }
+
+  static createForTesting = (): StorageCache => new StorageCache(newDefaultStorage())
+
+  static create = async (storage: Storage): Promise<StorageCache> => {
+    const storageCache = new StorageCache(storage)
+    await storageCache.migrate()
+    return storageCache
+  }
+
+  migrate = async (): Promise<void> => {
+    await migrateToNewFavoriteFormat(this)
+  }
+
+  /**
+   * Returns a storage item for the given key.
+   * The item should be treated as immutable and may not be modified.
+   *
+   * @param key The key of the storage item
+   */
+  getItem = <T extends keyof Storage>(key: T): Readonly<Storage[T]> => this.storage[key]
+
+  /**
+   * Returns a storage item that may be modified
+   *
+   * @param key The key of the storage item
+   */
+  getMutableItem = <T extends keyof Storage>(key: T): Storage[T] => JSON.parse(JSON.stringify(this.getItem(key)))
```

Comment:
🤔 I'm not sure the `getMutableItem` is necessary on top of `getItem` with `setItem`. But that might be a bit of React instinct in me, to only have one-way data binding.

Reply:
Aaaah, I see, it's really more of a copy of the item that we get there, which is why that result is mutable but the actual item is not. Got it, thank you! Yeah, I don't see a more elegant solution either :(